METRICS-1302: Added prefix support for resource labels.

apoorvmittal10 opened the pull request:
- Added `configuration` to override `metric dimension` name.
- Added `prefix` support for `metric` and `resource` labels.
- Default `resourceLabelPrefix`  to `resource`.
- Ingestion Spec:
```
        "type": "opencensus-protobuf",
        "metricDimension": "metric.dimension",
        "metricLabelPrefix": "des.",
        "resourceLabelPrefix": "res.", 
```

- Added `test cases` for the change:
   - Test `Default` behaviour.
   - Test with `custom` prefix.
   - Test with `no prefix`.

sirianni commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -191,4 +197,10 @@ private void addDerivedMetricsRow(Map<String, Object> derivedMetrics, List<Strin
+  private String resourceLabelsKeyMapper(String key)
+  {
+    return Strings.isNullOrEmpty(config.getResourceLabelsPrefix()) ? key :
+        config.getResourceLabelsPrefix() + "." + key;
```
It would be more flexible to not hardcode the `.` delimiter in the code and instead have it be part of the prefix itself.  This would allow users with other naming conventions to use those (e.g. delimit by `_`, etc.).

apoorvmittal10 replied:
I made the `delimiter` as part of `prefix` itself. I was thinking whether it makes sense to have another config for `delimiter`.

sirianni commented on extensions-contrib/opencensus-extensions/src/test/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParserTest.java:
```diff
@@ -93,6 +96,8 @@ public void setUp()
         ), null
     );
+
+    config = Mockito.mock(OpenCensusProtobufConfig.class);
```
Although I do love Mockito, is mocking really necessary for the `OpenCensusProtobufConfig` class?  It's just a simple data class.  Why not just expose a setter for the `resourceLabelsPrefix` field?

apoorvmittal10 replied:
I always try not to modify actual code just for testing as we can achieve behaviour by `mocking`. 

Though I have added an instance creator method, as in future there might be other configs hence it will be convenient to add required test config in the `getConfig` method itself.

sirianni replied:
OK fair.

xvrl replied:
agree with @sirianni mocking seems unnecessary. @apoorvmittal10 if you have control over a class, or are writing a new class, it's usually better to structure it so you can test the actual behavior. Otherwise you're only be testing your mocked behavior, which may not reflect reality.

sirianni commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufConfig.java:
```diff
@@ -0,0 +1,33 @@
+public class OpenCensusProtobufConfig
+{
+  @JsonProperty
+  private String resourceLabelsPrefix = "resource";
```
Does the plural `labels` make sense here?  Would `resourceLabelPrefix` be better?

apoorvmittal10 replied:
Done.

xvrl replied:
If we want to add a configuration for the metric name dimension, then it would probably be called `metricDimension`, in which case to be consistent we should probably call this `resourceDimensionPrefix`?

apoorvmittal10 replied:
Done.

sirianni commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufExtensionsModule.java:
```diff
@@ -45,5 +46,6 @@
   @Override
   public void configure(Binder binder)
   {
+    JsonConfigProvider.bind(binder, "druid.opencensus.protobuf", OpenCensusProtobufConfig.class);
```
Is `protobuf` necessary here?  Use of OpenCensus kind of implies protobuf.  And even if it didn't wouldn't processing of resource labels be format-agnostic (protobuf or json)?

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufExtensionsModule.java:
```diff
@@ -45,5 +46,6 @@
   @Override
   public void configure(Binder binder)
   {
+    JsonConfigProvider.bind(binder, "druid.opencensus", OpenCensusProtobufConfig.class);
```
I don't think it makes sense for this to be a service-level configuration. Instead, this should be configurable as part off the ingestion spec.

apoorvmittal10 replied:
@xvrl I looked into the `configuration` of different `extensions` and saw the configuration is always at service-level.

I now also checked `ingestion spec` and it has `data schema`, io config`, `tuning config`. Can you please help me regarding an example extending `ingestion spec` and how the prefix configuration in our case should go and where.

https://druid.apache.org/docs/latest/development/extensions.html
https://druid.apache.org/docs/latest/development/extensions-core/google.html

xvrl replied:
@apoorvmittal10 there are many different types of extensions. Unlike parsers, many extensions cannot be configured at runtime. Parser configurations are defined as part of the constructor, so all you need to make the parser configurable is to add a constructor argument ` @JsonProperty("resourceDimensionPrefix") String resourcePrefix` to `OpenCensusProtobufInputRowParser`

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -116,7 +121,8 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
       // Add common resourceLabels.
-      Map<String, Object> labels = new HashMap<>(metric.getResource().getLabelsMap());
+      Map<String, Object> labels = new HashMap<>(metric.getResource().getLabelsMap().entrySet().stream()
+          .collect(Collectors.toMap(entry -> resourceLabelsKeyMapper(entry.getKey()), Map.Entry::getValue)));
```
it seems unnecessary to prefix the labels a second time when we've already done so when updating `recordDimensions`.

apoorvmittal10 replied:
A complete miss by me, thanks for pointing out.

apoorvmittal10 replied:
Optimized further with single key mapper for each type, processed once and used either the `key sets` or `label keys list`.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufConfig.java:
```diff
@@ -0,0 +1,43 @@
+{
+
+  @JsonProperty
+  private String resourceLabelPrefix = "resource.";
```
while we're at it, we might as well add a prefix config for metric labels as well.
It would also be useful to configure the "metric name" dimensions as well. Our current choice of `name` is probably not the greatest and it could potentially clash with metric labels.

apoorvmittal10 replied:
@xvrl Do you mean similar prefix behaviour for `metric.getMetricDescriptor().getLabelKeys`?

xvrl replied:
yes, two things:
- we'd like to have a prefix for metric.getMetricDescriptor().getLabelKeys(), which can be empty by default.
- we'd like to allow the user to override the `NAME` constant for the metric name

apoorvmittal10 replied:
Done.

xvrl commented:
please target the 0.17.0-confluent branch, since that's the only one we currently run tests for.

apoorvmittal10 commented:
> please target the 0.17.0-confluent branch, since that's the only one we currently run tests for.

@xvrl Shall change the branch with addressed changes but I have some minor doubts, please if you can help.

apoorvmittal10 commented:
@xvrl I have addressed review comments, moved config to ingestion spec.
Optimized code further to process the `key mapper` once for each `prefix type`.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -88,17 +103,29 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
 
+    // Fetch metric name.
+    String metricName = !Strings.isNullOrEmpty(this.metricDimension) ? this.metricDimension :
+        metric.getMetricDescriptor().getName();
```
this should be done at initialization time in the constructor

apoorvmittal10 replied:
I think we can just have a `boolean` replacing `NullorEmpty` method can be taken in constructor.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +49,29 @@
-      @JsonProperty("parseSpec") ParseSpec parseSpec
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("metricDimension") String metricDimension,
+      @JsonProperty("descriptorDimensionPrefix") String descriptorPrefix,
```
I'm not sure if `descriptor` makes it very obvious this is referring to metric labels. I was wrong here, `metricLabelPrefix` and `resourceLabelPrefix` will probably be easier to understand.

apoorvmittal10 replied:
Done.

xvrl replied:
I don't see this change

apoorvmittal10 replied:
Missed the commit. Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -191,4 +218,9 @@ private void addDerivedMetricsRow(Map<String, Object> derivedMetrics, List<Strin
 
+  private String prefixKeyMapper(String prefix, String key)
+  {
+    return Strings.isNullOrEmpty(prefix) ? key : prefix + key;
```
same here, let's not do the null checks every time, let's resolve the prefix in the constructor when we initialize the parser.

apoorvmittal10 replied:
Moved `NullOrEmpty` method as one time check in constructor with boolean variable. I think that can be only optimization we can have here.

apoorvmittal10 replied:
Done.

apoorvmittal10 commented:
@xvrl Gentle reminder for review.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +49,36 @@
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("metricDimension") String metricDimension,
+      @JsonProperty("descriptorDimensionPrefix") String metricPrefix,
+      @JsonProperty("resourceDimensionPrefix") String resourcePrefix
```
I don't see the change we had discussed here https://github.com/confluentinc/druid/pull/14#discussion_r407660090

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -88,17 +110,28 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
     }
 
+    // Fetch metric name.
+    String metricName = this.hasMetricDimension ? this.metricDimension : metric.getMetricDescriptor().getName();
```
I think there was a misunderstanding here, we are not trying to change the name of the metric itself. The point of having `metricDimension` was to change the value of the *dimension*, i.e. the `NAME` constant defined here https://github.com/confluentinc/druid/pull/14/files#diff-13a54faa5722482b59f3f83c9fdfd4deR52

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +49,36 @@
+    this.resourceLabelPrefix = resourcePrefix == null ? DEFAULT_RESOURCE_PREFIX : resourcePrefix;
+
+    this.hasMetricDimension = !Strings.isNullOrEmpty(this.metricDimension);
+    this.hasMetricLabelPrefix = !Strings.isNullOrEmpty(this.metricLabelPrefix);
```
I don't think we need those variables, you can achieve the same behavior by setting `this.metricLabelPrefix = Strings.nullToEmpty(metricLabelPrefix)`

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -88,17 +110,28 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
+
+    // Process metric descriptor labels map keys.
+    List<String> descriptorLabels = metric.getMetricDescriptor().getLabelKeysList().stream()
+        .map(s -> (!this.hasMetricLabelPrefix ? s.getKey() : this.metricLabelPrefix + s.getKey()))
```
as mentioned above if we set the prefix to empty by default if not set, we can avoid the complexity here.

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +51,31 @@
     this.parseSpec = parseSpec;
     this.dimensions = parseSpec.getDimensionsSpec().getDimensionNames();
+    this.metricDimension = !Strings.isNullOrEmpty(metricDimension) ? metricDimension : DEFAULT_METRIC_DIMENSION;
+    this.metricLabelPrefix = NullHandling.nullToEmptyIfNeeded(metricPrefix);
```
why would we be using `NullHandling` here instead of  `Strings.nullToEmpty()`? `NullHandling` is only needed to manipulate actual values, but not for dimensions names.

apoorvmittal10 replied:
Druid code checks doesn't allow to use `Strings.nullToEmpty()`.
https://github.com/confluentinc/druid/blob/d67772c02ce230c3be339d8a1ce336405d89381c/codestyle/checkstyle.xml#L205

xvrl replied:
that's to avoid possible bugs with null handling in values, you can override that check in this particular case, or use `StringUtils.nullToEmptyNonDruidDataString` from `org.apache.druid.java.util.common.StringUtils`

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +51,31 @@
   {
     this.parseSpec = parseSpec;
     this.dimensions = parseSpec.getDimensionsSpec().getDimensionNames();
+    this.metricDimension = !Strings.isNullOrEmpty(metricDimension) ? metricDimension : DEFAULT_METRIC_DIMENSION;
```
This ternary expression seems a bit convoluted. Can we remove the negation and flip the expressions to make it easier to read?

apoorvmittal10 replied:
Done.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +51,31 @@
     this.dimensions = parseSpec.getDimensionsSpec().getDimensionNames();
+    this.metricDimension = !Strings.isNullOrEmpty(metricDimension) ? metricDimension : DEFAULT_METRIC_DIMENSION;
+    this.metricLabelPrefix = NullHandling.nullToEmptyIfNeeded(metricPrefix);
+    this.resourceLabelPrefix = resourcePrefix != null ? resourcePrefix : DEFAULT_RESOURCE_PREFIX;
```
why are we handling `resourceLabelPrefix` and `metricLabelPrefix` differently? They should both be treated the same way. What we do here seems perfectly fine for metricsLabelPrefix as well.

apoorvmittal10 replied:
As per the discussion we had I think we agreed to have `resource.` as default prefix for `resource` labels where `metric` labels will default to `empty`.

https://github.com/confluentinc/druid/pull/14#discussion_r406329904

xvrl replied:
right, we could do this by explicitly having `DEFAULT_METRIC_PREFIX = ""` to make it explicit what the defaults are, but if you want to leave it as is that's fine too.

apoorvmittal10 replied:
Leaving it as it is for now. Will change if needed for some other default prefix.

xvrl replied:
that's fine, it was mostly a suggestion to make the pattern more obvious to someone reading the code.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -88,22 +107,30 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
+        .collect(Collectors.toList());
+
+    // Process resource labels map.
+    Map<String, Object> resourceLabelsMap = new HashMap<>(metric.getResource().getLabelsMap().entrySet().stream()
```
is the `new HashMap<>()` necessary here? doesn't the `Collectors.toMap` already create the map?

apoorvmittal10 replied:
Removed, seems like I am missing important details. Will be more stringent in future, apologies.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -88,22 +107,30 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
+      // MetricDimension, VALUE dimensions will not be present in labelKeysList or Metric.Resource
+      // map as they are derived dimensions, which get populated while parsing data for timeSeries
+      // hence add them to recordDimensions.
+      recordDimensions.add(this.metricDimension);
```
nit, the `this.` prefix is not necessary here

apoorvmittal10 replied:
I find `this.` increases readability i.e. distinguish easily between local vs global variables hence leaving it as it is.

xvrl replied:
we can leave as is, but the convention in the Druid codebase we typically only use this pattern when there is ambiguity between local and member variables, e.g. it often happens in the constructor. It can make the code harder to read when the code often refers to member variables, in which case it becomes unnecessarily verbose.
This PR doesn't seem consistent about it either (see https://github.com/confluentinc/druid/pull/14/files#diff-13a54faa5722482b59f3f83c9fdfd4deR93-R95 which doesn't use it)

apoorvmittal10 replied:
Point Taken, updated the code.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -48,19 +51,31 @@
 
+  private final String metricDimension;
+  private final String metricLabelPrefix;
+  private final String resourceLabelPrefix;
```
I just realized you didn't add any @UsonProperty annotated getters for those properties, which means they will be left out when serializing to json. This is required for nodes to communicate the parser spec to each-other

apoorvmittal10 replied:
@xvrl Created PR: https://github.com/confluentinc/druid/pull/15
Added additional test cases and `labels` for `distribution` metric too.
